File: atividades/exercicios_fixacao/aula4_sobreescrita.py
class Funcionario:
    def __init__(self, nome, salario_base):
        self.nome = nome 
        self.salario_base = salario_base

    def calcular_bonus(self):
        bonus = 0.05 * self.salario_base
        self.salario_base = self.salario_base + bonus

        print(f"{self.nome}, seu bonus é de: R${bonus} | O salario final é R${self.salario_base}")

class Gerente(Funcionario):
    def __init__(self, nome, salario_base, valor_fixo):
        super().__init__(nome, salario_base)
        self.valor_fixo = valor_fixo

    def calcular_bonus(self):
        super().calcular_bonus()
        self.salario_base = self.salario_base + 1000

        print(f"O Gerente - {self.nome} - vai ganhar com o bonus de valor fixo (R${self.valor_fixo}) R${self.salario_base}")

class Vendedor(Funcionario):
    def __init__(self, nome, salario_base):
        super().__init__(nome, salario_base)

    def calcular_bonus(self):
        bonus = 0.1 * self.salario_base
        self.salario_base = self.salario_base + bonus

        print(f"{self.nome}, seu bonus é de: R${bonus} | O salario final é R${self.salario_base}")

funcionario = Funcionario("Lucas", 40000)
gerente = Gerente("Xucas", 50000, 1000)
vendedor = Vendedor("Luxas", 5000)

funcionario.calcular_bonus()
gerente.calcular_bonus()
vendedor.calcular_bonus()

File: atividades/exercicios_fixacao/Aula_4/C++/aula4_sobreescrita.cpp
#include <iostream>
#include <string>
#include <memory>
#include <iomanip>

class Funcionario {
protected:
    std::string nome;
    double salario_base;

public:
    Funcionario (std::string n, double s) 
        : nome(n), salario_base(s) {}
    
    virtual void calcular_bonus() {
        double bonus = salario_base * 0.05; // Aplica 5% sobre o salario
        salario_base = salario_base + bonus;

        std::cout << nome 
                  << ", seu bonus e de: R$ " << std::fixed << std::setprecision(2) << bonus
                  << " | O salario final e R$ " << std::fixed << std::setprecision(2) << salario_base << "\n";
    }   

    virtual ~Funcionario() = default;
};

class Gerente : public Funcionario {
private:
    double valor_fixo;

public:
    Gerente (std::string n, double s, double vf) 
        : Funcionario(n, s), valor_fixo(vf) {}

    void calcular_bonus() override {
        Funcionario::calcular_bonus();

        salario_base = salario_base + 1000;
        std::cout << "O Gerente: " << nome
                  << " vai ganhar com o bonus de valor fixo (R$ " << std::fixed << std::setprecision(2) << valor_fixo << ")"
                  << "R$ " << std::fixed << std::setprecision(2) << salario_base << "\n";
    }
};

class Vendedor : public Funcionario {
public:
    Vendedor (std::string n, double s) 
        : Funcionario(n, s) {}

    void calcular_bonus() override {
        double bonus = salario_base * 0.10; // Aplica 10% sobre o salario
        salario_base = salario_base + bonus;

        std::cout << "O vendedor " << nome 
                  << ", seu bonus e de: R$ " << std::fixed << std::setprecision(2) << bonus
                  << " | O salario final e R$ " << std::fixed << std::setprecision(2) << salario_base << "\n";
    }
};

int main() {
    std::unique_ptr<Funcionario> f1 = std::make_unique<Funcionario>("Lucas", 30000);
    std::unique_ptr<Funcionario> g1 = std::make_unique<Gerente>("Xucas", 40000, 1000);
    std::unique_ptr<Funcionario> v1 = std::make_unique<Vendedor>("Luxas", 50000);

    f1->calcular_bonus();
    g1->calcular_bonus();
    v1->calcular_bonus();
    return 0;
}